lib/useOnyx.ts

```diff
@@ -140,16 +140,16 @@ function useOnyx<TKey extends OnyxKey, TReturnValue = OnyxValue<TKey>>(key: TKey
 
     const subscribe = useCallback(
         (onStoreChange: () => void) => {
-            connectionIDRef.current = Onyx.connect({
-                key: key as CollectionKeyBase,
+            connectionIDRef.current = Onyx.connect<CollectionKeyBase>({
+                key,
                 callback: () => {
                     // We don't need to update the Onyx cache again here, when `callback` is called the cache is already
                     // expected to be updated, so we just signal that the store changed and `getSnapshot()` can be called again.
                     isFirstConnectionRef.current = false;
                     onStoreChange();
                 },
                 initWithStoredValues: options?.initWithStoredValues,
-                waitForCollectionCallback: OnyxUtils.isCollectionKey(key),
+                waitForCollectionCallback: OnyxUtils.isCollectionKey(key) as true,
```

Comment:
I don't think these types are well defined. Why are we using the case `as true`? cast should be avoided, why do we need one here? If we are using `CollectionKeyBase` above, shouldn't we just put `waitForCollectionCallback: true` instead of calling `OnyxUtils.isCollectionKey(key)`? If this may not be a collection, then why did we hardcode `CollectionKeyBase` up there in the first place?

Reply:
It is a standard practice to use assertion in library code, TS isn't ideal and sometimes we have to use either assertions or ts-expect-error comment.

Here it is necessary to use an assertion due to how connect method is defined:
```ts
function connect<TKey extends OnyxKey>(connectOptions: ConnectOptions<TKey>): number {

👇 

type ConnectOptions<TKey extends OnyxKey> = (CollectionConnectOptions<TKey> | DefaultConnectOptions<TKey>) & (BaseConnectOptions | WithOnyxConnectOptions<TKey>);

👇 

type DefaultConnectOptions<TKey extends OnyxKey> = {
    key: TKey;
    callback?: DefaultConnectCallback<TKey>;
    waitForCollectionCallback?: false;
};

/** Represents the callback function used in `Onyx.connect()` method with a collection key. */
type CollectionConnectOptions<TKey extends OnyxKey> = {
    key: TKey extends CollectionKeyBase ? TKey : never;
    callback?: CollectionConnectCallback<TKey>;
    waitForCollectionCallback: true;
};
```

Because we use `waitForCollectionCallback`, Typescript assumes that the key is `CollectionKeyBase` 👉 `waitForCollectionCallback` has to be true in this case. I tried different approaches to this but wasn't able to do it without assertion.

The only idea that comes to my mind is to change the types in `DefaultConnectOptions` and `CollectionConnectOptions` so that waitForCollectionCallback is boolean in both cases:
```ts
/** Represents the callback function used in `Onyx.connect()` method with a regular key. */
type DefaultConnectOptions<TKey extends OnyxKey> = {
    key: TKey;
    callback?: DefaultConnectCallback<TKey>;
    waitForCollectionCallback?: boolean;
};

/** Represents the callback function used in `Onyx.connect()` method with a collection key. */
type CollectionConnectOptions<TKey extends OnyxKey> = {
    key: TKey extends CollectionKeyBase ? TKey : never;
    callback?: CollectionConnectCallback<TKey>;
    waitForCollectionCallback?: boolean;
};
```

Reply:
I appreciate the explanation, and I'm not sure really if that would be better. I tried playing around adding a super type for `DefaultConnectOptions` and `CollectionConnectOptions`, but couldn't make it work as I wanted.

I think the types that you defined will work well when we use `Onyx.connect` from the App, so the assertion needed here is a minor detail.

Reply:
Also, I learned here that in typescript this is called `assertion` as you say and not a `cast` as a I said before :P

Reply:
Interesting, isn't cast and assertion the same thing? 😮

Reply:
> Interesting, isn't cast and assertion the same thing? 😮

Honestly, not sure, maybe it depends on the language? You made me doubt because in languages like C++ / java, casting happens at runtime and may fail, here it is different since these types is only used when building and since you called it `assertion`, I thought I was wrong about calling it a `cast`. 🤷